### PR TITLE
Move logic for generating URLs for the view, edit and source links of catalog entities from the catalog frontend into the backend

### PR DESCRIPTION
**Pull request by Fox32**
As discussed in #4575, this PR moves the generation of the fallback URLs to the backend by populating the corresponding annotations.

This also extends the integration package to provide an API to retrieve edit URL for each SCM system. As I don't know much about Azure, I skipped that.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: .changeset/lucky-schools-bathe.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/integration': patch
+---
+
+Add `resolveEditUrl` to integrations to resolve an URL that can be used to edit
```
Comment (adamdmharvey) [inline .changeset/lucky-schools-bathe.md:5]:
```suggestion
Add `resolveEditUrl` to integrations to resolve a URL that can be used to edit
```

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts [hunk 2]
```diff
@@ -15,31 +15,79 @@
...
+  // TODO: Resolve URL from integrations doesn't actually support resolving to folders...
+  // In github one would have to switch from blob to tree urls...
+  // I wonder if this is something I should put into integrations.
+
+  const obj = gitUrlParse(url);
+
+  obj.filepath = '/';
+
+  return gitUrlParse.stringify(obj);
+}
```
Comment (Fox32) [inline plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:93]: @freben I thought it would be a good idea if the "View Source" link navigates the user to the actual repository:

image

I first thought that I can use resolveUrl from integration, but resolving to '/' doesn't actually work for GitHub has they use different URLs for viewing a folder. Alternative is to build an URL to the repository instead, or use this workaround. Not sure if that works for other providers. We could also move this to the integration package.

What do you think?

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4702-4vvwyhnxier0tcya9kjxu0qolpkkgky4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037e218f285afe8e1989f10

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts [hunk 3]
```diff
@@ -15,31 +15,79 @@
...
-      entity,
-    );
+    const { integrations } = this.options;
+
+    const annotations: Record<string, string> = {
+      [LOCATION_ANNOTATION]: `${location.type}:${location.target}`,
+      [ORIGIN_LOCATION_ANNOTATION]: `${originLocation.type}:${originLocation.target}`,
+    };
+
+    if (isValidUrl(location.target)) {
```
Comment (Fox32) [inline plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:51]: Guard on location type?

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts (hunk 2 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:93]: These url dissection things are exactly the kind of stuff that fits well there!

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:93]: Would be cool to extend what we already have to support resolving `/` effectively, if that's useful 

https://github.com/backstage/backstage/blob/master/packages/integration/src/ScmIntegrations.ts#L86

Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:93]: It would perfectly fit `resolveUrl`, but I don't think that it is possible. To generate a valid GitHub url I would have to know upfront whether I link to a folder or a file.

For example:

```
https://github.com/backstage/backstage/blob/master/packages/catalog-model/README.md
```

Can be resolved via `/` to: 

```
https://github.com/backstage/backstage/blob/master/
```

But that link is wrong, because it doesn't display the folder contents. It should be (note the `tree`):

```
https://github.com/backstage/backstage/tree/master/
```

But I don't think there is a way to decide whether a path is a file or a directory. What I could do, is to pass another argument that specifies that I expect the result to be a directory?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:93]: Myeah I dunno. Using `tree` pointing to a file actually redirects to `blob` :P but yeah some of this functionality would actually require using the appropriate client lib instead of raw url manipulation to be safe

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts [hunk 4]
```diff
@@ -15,31 +15,79 @@
...
+          : undefined;
+
+      annotations[VIEW_URL_ANNOTATION] = location.target;
+      const editUrl = scmIntegration?.resolveEditUrl(location.target);
+
+      if (editUrl) {
+        annotations[EDIT_URL_ANNOTATION] = editUrl;
+      }
+
+      annotations[SOURCE_LOCATION_ANNOTATION] = extractRepositoryRoot(
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:64]: Thinking about this, I don't think it's right after all. You probably want to do something like "if the last path segment matches `/\.ya?ml$/`, then strip it off". Because in a monorepo situation where the file is not at the root of the repo but rather together with the module that it describes, you want to go the folder that the file is in.

Does `scmIntegration.resolveUrl({ base: location.target, url: './' })` or something like that work for that? Can't check right now :)

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:64]: Oh right, the thing with `/tree/` vs `/blob/`. Yeah well, github at least seems to redirect properly both ways, so clicking it will actually work in practice

Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:64]: Yes in general `./` would be better than `/`, especially in Mono Repositories.

I will play around a bit with how that would work with GitLab and Bitbucket.

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts (hunk 2 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:93]: GitLab has a redirect between both tree and blob which is nice. Bitbucket is even simpler and doesn't make a difference 🙂 

Not sure about azure, but neither could I implement the edit URL there because I miss access. Or do they have public repositories I can play around with?

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4702-4vvwyhnxier0tcya9kjxu0qolpkkgky4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037e218f285afe8e1989f10

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts (hunk 2 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts]: Hm not sure. But maybe just make it good enough for now (perhaps even just return the location unchanged) and let it evolve over time

Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts]: Good, that is the way I went 🙂 I think I also left enough comments there for the next one to finish the Azure part

File: packages/integration/src/github/GitHubIntegration.ts [hunk 5]
```diff
@@ -47,6 +47,13 @@ export class GitHubIntegration implements ScmIntegration {
...
   resolveUrl(options: { url: string; base: string }): string {
-    return defaultScmResolveUrl(options);
+    // GitHub uses blob URLs for files and tree urls for directory listings. But
+    // there is a redirect from tree to blob for files, so we can always return
+    // tree urls here.
+    return defaultScmResolveUrl(options).replace('/blob/', '/tree/');
+  }
+
+  resolveEditUrl(url: string): string {
+    return url.replace('/blob/', '/edit/');
```
Comment (Rugvip) [inline packages/integration/src/github/GitHubIntegration.ts:57]: What about `https://github.com/webmodules/blob/blob/master/README.md` tho? Tbh I think it's worth handling that edge-case xP, at least we don't need to support `blob/blob/blob` branches

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts [hunk 6]
```diff
@@ -52,8 +57,110 @@ describe('AnnotateLocationEntityProcessor', () => {
...
+    it('does not override existing annotations', async () => {
+      const entity: Entity = {
+        apiVersion: 'backstage.io/v1alpha1',
+        kind: 'Component',
+        metadata: {
+          name: 'my-component',
+          annotations: {
+            'backstage.io/view-url': 'https://example.com/view',
+            'backstage.io/edit-url': 'https://example.com/edit',
+            'backstage.io/source-location': 'https://example.com/source',
```
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: `backstage.io/source-location` uses the catalog location format, i.e. `url: https://example.com/source`, noticed that's not assumed throughout this PR 😅 

Kinda wanna challenge that a lil bit though @freben, what kind of non-url location would actually be useful? I'm assuming stuff like GitHub orgs would not have a `source-location` as they're not source code. For org-specific custom locations we can't handle those here anyway, so is there actually any value to including the location type? With that I'm ofc also arguing that it would be `source-url` :p

**Review by Rugvip — COMMENTED** [review]
Yep, makes sense 👍 

Just gotta open up a can of worms xP

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts (hunk 6 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: I was thinking things like `sysmodel:<id>`. But besides, whether it happens to be a url normally, it is location like, isn't it? And other locations have a type and target. I was intending for this field to be consumable by processors / general code. But you mean that it'd be descoped to be a browser-only thing for human consumption then right?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: Doesn't it make a lot of sense to have the ability for tooling to e.g. traverse to the pointrd-at repo and list the number of unmerged pull requests etc?

Bear in mind - this annotation is essentially the _replacement_ for managed-by-location when what you point to isn't where the actual yaml file is

Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: It wouldn't be something like `sysmodel:<id>` tho right? that's be the manage-by-location. `source-location` is strictly the location of the source code, and doesn't really have that much to do with locations right?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: I meant that managed-by-location is strictly "where this yaml was sourced from" and source-location is "what i am describing" which can be two different things

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: In the sysmodel example they could for example be `sysmodel:*` and sysmodel:<id>`, respectively

Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: So we got `backstage.io/managed-by-origin-location`, `backstage.io/managed-by-location`, and `backstage.io/source-location`, all referring to the source of the entity in different ways, at different levels? Perhaps a `location-ancestry` makes sense to bring it back down to just 2 annotations? With a list of all locations that were used to track down the entity.

Or am I missing the purpose of the 3 annotations?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: Yeah one for the actual registrered location (to keep track of which ones to unregister if you remove the location), one to do a single step of ancestry (can make several queries to stitch the full picture together), and one to point out the actual thing being described.

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: Should have called them `registered-location`, `catalog-info-yaml-location` and `source-code-location` ;)

Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: So back to the original question, when is `source-code-location` not a URL?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: I guess we don't know yet. When the source is in a not-browser-based interface I guess. But isn't that the same as asking when will `backstage.io/managed-by-origin-location` not be a URL? 🤔 Maybe not

File: packages/integration/src/github/GitHubIntegration.ts (hunk 5 above)
Comment (Fox32) [reply packages/integration/src/github/GitHubIntegration.ts:57]: Hmm I guess that is an issue we should fix. Should be easy to fix with a simple regex, give me a second! Insert comic here that starts with this sentence but ends in existential dread.

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts (hunk 6 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts:84]: So right now the [documentation](https://backstage.io/docs/features/software-catalog/well-known-annotations#backstageiosource-location) says that it's a 

> `Location` reference

so it would be correct to include `url:` and have the UI parse such location references. The UI would only work for `url:` type though. We might consider to add a small parser for the format to `catalog-model` as we already have multiple of them.

File: packages/integration/src/github/GitHubIntegration.ts (hunk 5 above)
Comment (Fox32) [reply packages/integration/src/github/GitHubIntegration.ts]: Done

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4702-4vvwyhnxier0tcya9kjxu0qolpkkgky4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037e218f285afe8e1989f10

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4702-4vvwyhnxier0tcya9kjxu0qolpkkgky4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037e218f285afe8e1989f10

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts (hunk 6 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts]: Done, can you have another look?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts]: +1 been meaning to add those for quite some time, so i did :) https://github.com/backstage/backstage/pull/4835

Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts]: Yep, didn't mean to block this PR on the discussion, but it did need to be in sync with the schema

Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.test.ts]: I will wait for #4835 and remove the duplicate changes on monday, I guess than we are ready to go

**Comment by Fox32** [comment]
Rebased on #4835 and adjusted my code to the latest changes. Should be ready now.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4702-4vvwyhnxier0tcya9kjxu0qolpkkgky4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037e218f285afe8e1989f10

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts [hunk 7]
```diff
@@ -15,31 +15,71 @@
...
+            {
+              [LOCATION_ANNOTATION]: stringifyLocationReference(location),
+              [ORIGIN_LOCATION_ANNOTATION]: stringifyLocationReference(
+                originLocation,
+              ),
+              [VIEW_URL_ANNOTATION]: viewUrl,
+              [EDIT_URL_ANNOTATION]: editUrl,
+              [SOURCE_LOCATION_ANNOTATION]: sourceLocation,
+            },
+            identity,
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:81]: do you need the pick and identity? isn't the merge a deep merge anyway? 🤔

File: plugins/catalog/src/utils/getEntitySourceLocation.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  const sourceLocation =
+    entity.metadata.annotations?.[SOURCE_LOCATION_ANNOTATION];
+
+  if (!sourceLocation) {
+    return undefined;
+  }
+
+  try {
+    const sourceLocationRef = parseLocationReference(sourceLocation);
+    const scmIntegrations = ScmIntegrations.fromConfig(config);
```
Comment (freben) [inline plugins/catalog/src/utils/getEntitySourceLocation.ts:43]: This makes me realise - i wonder if this constructor actually is robust for all providers _on the frontend_, or if it will reject things (perhaps even throw) as it does not find any tokens / passwords etc that are only available in the backend because of the config system stripping them out on the frontend...

Also nothing that this is pretty unfortunate to have to construct over and over especially in a large table or so. It should be available as `useIntegrations` basically and passed in by the component. But OK, let's do that separately.

File: plugins/catalog/src/utils/getEntitySourceLocation.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+} from '@backstage/catalog-model';
+import { ConfigApi } from '@backstage/core';
+import { ScmIntegrations } from '@backstage/integration';
+
+export type EntitySourceLocation = {
+  url: string;
+  type?: string;
+};
+
+export function getEntitySourceLocation(
```
Comment (freben) [inline plugins/catalog/src/utils/getEntitySourceLocation.ts:30]: hm so ... this isn't really for external use right? since it's in this package

just a slight confusion on location type+target vs this url+type where the type isn't a location type :) maybe call them `locationTargetUrl` and `integrationType` to make it clearer

**Review by freben — APPROVED** [review]
Little nits, but +1 from me!

**Review by OrkoHunter — APPROVED** [review]

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts (hunk 7 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:81]: It's to avoid having keys that are filled with `undefined`, as `editUrl`, `viewUrl` and `sourceLocation` might be `undefined`. I think my experiments showed that merge would copy undefined values.

File: plugins/catalog/src/utils/getEntitySourceLocation.ts (hunk 8 above)
Comment (Fox32) [reply plugins/catalog/src/utils/getEntitySourceLocation.ts:43]: Hmm, yeah. Having that cached would be neat.

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts (hunk 7 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:81]: Ingestion would that fail with:

```
[1] 2021-03-08T10:59:29.090Z catalog warn Failed item in location file:../../plugins/scaffolder-backend/sample-templates/all-templates.yaml, InputError: Malformed envelope, /metadata/annotations/backstage.io~1view-url should be string
```

File: plugins/catalog/src/utils/getEntitySourceLocation.ts (hunk 9 above)
Comment (Fox32) [reply plugins/catalog/src/utils/getEntitySourceLocation.ts:30]: No, this is not exported from the catalog package, just used internally. I like the naming suggestion, updated!

File: plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts (hunk 7 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts:81]: Ah yes that makes sense! Yeah I have code somewhere else that does like `stripKeysWithUndefinedValues` for the same reason

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4702-4vvwyhnxier0tcya9kjxu0qolpkkgky4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6037e218f285afe8e1989f10
